### PR TITLE
Mesh Generator Directed Acyclic Graph Cleanup

### PR DESCRIPTION
**Pull request by permcody**
Add optional parameter to select final generator
Throw error where the final mesh is not deterministic

New tests and doco added

Refs #14058

### CODE REVIEW

**Comment by dschwen** [comment]
DAG?

**Comment by permcody** [comment]
@friedmud - This is ready for review. BISON has been patched.

File: framework/doc/content/syntax/Mesh/index.md [hunk 1]
```diff
@@ -23,6 +23,24 @@ of Mesh objects refer to the individual object pages listed below.
...
+from a serious of points and connectivity, adding features on the fly, linearly transforming the mesh, stitching
+together pieces of meshes, etc. There are several built-in generators but this system is also extendable. MeshGenerators
+may or may not consumer the output from other generators and produce a single mesh. They can be chained together
+through dependencies so that complex meshes may be built up from a series of simple processes.
+
+### DAG and final mesh selection
+
+When chaining together several MeshGenerators, you are implicitly creating a DAG (directed acyclic graph).
+MOOSE evaluates and generates the individual objects to build up your final mesh. If your input file has
+multiple end points, (e.g. B->A and C->A) then MOOSE will issue an error and terminate. Generally, it doesn't m
```
Comment (sapitts) [inline framework/doc/content/syntax/Mesh/index.md]: Extra m at the end of the line hiding here

Comment (permcody) [reply framework/doc/content/syntax/Mesh/index.md]: Dammnit!

**Comment by permcody** [comment]
So I started thinking about this randomly on the weekend when I was with my boys and realized that the check I put here isn't sufficient. The DependencyResolver only tells you the order when you can run thing. It wasn't designed to tell that the DAG had a single end point. I thought of several cases which would pass this check but would fail to produce a single final mesh. More work is needed here...

File: framework/src/actions/SetupMeshAction.C [hunk 2]
```diff
@@ -236,10 +236,10 @@ SetupMeshAction::act()
           _type = "MeshGeneratorMesh";
           auto original_params = _moose_object_pars;
           _moose_object_pars = _factory.getValidParams("MeshGeneratorMesh");
-          std::vector<std::string> params_to_apply{"parallel_type", "patch_size", "uniform_refine"};
```
Comment (permcody) [inline framework/src/actions/SetupMeshAction.C]: This line of code is asking for trouble. I think it's best to just extract all of the parameters for the new object if it's programmatically changed.

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**